### DISCUSSION TITLE
Transport throttling?

### DISCUSSION THREAD

**Discussion by jeffj-archer** [OpenDDS/OpenDDS#4555] [General]
We have an app that is using a relay.  We're doing RTPS_UDP.  Our publisher is sending a topic/message with ~200K payloads.  We see messages getting "stuck" on the subscriber side.  It's as if there's some throttling going on, because as other messages arrive it eventually pushes prior published messages out (the subscriber finally gets called-back and can consume the message).  Our app is latency sensitive, so these messages getting throttled until future messages push them out is problematic for us.  

We can play with message payload size and that seems to affect the underlying transport and how timely we're getting those messages on the subscriber side.  

We're looking for suggestions.  Other places we should be asking these questions?  Thank you!

-- jrj

**Comment by jrw972** [comment]
By "relay", I'm assuming the RtpsRelay that comes with OpenDDS.  If not, please clarify.

What is the QoS for the DataWriter and DataReader?
Are all samples in this topic 200K or is that exceptional?
What is the MTU in the network linking the participants and the relay?

In general, OpenDDS will invoke a listener or unblock a waitset as soon as possible.
So if you are perceiving something like throttling, there is some other issue.
Newer messages cannot displace an older message unless the QoS is set up for it, e.g., not reliable with finite history depth.

I'll make some assumptions to illustrate a possibility.
I'm not implying that this is what is happening but hopefully it will get you closer to a solution.

I'll assume that you are running over the public Internet with reliability.
Sources vary but a payload greater than 1500 bytes will be subject to IP fragmentation on the Internet.
So, a 200K sample will be fragmented at the IP level and sent.
There is a good chance that at least one of these messages will be lost.
Thus, the writer will send a heartbeat, the reader will negatively acknowledge, the writer will resend, and the process will repeat until you get lucky and all the fragmented packets arrive.
(Some routers don't support IP fragmentation and only forward the first fragment so IP fragmentation should generally be avoided.)
Obviously, this is going to increase your latency and you have already discovered that changing the payload size affects your results.
The solution, in this case, is to instruct OpenDDS to fragment samples to avoid IP fragmentation.
Most likely, you will still drop packets, however, any packet that is received doesn't need to be resent so the end result will be much better.
The configuration options for doing this are `SedpMaxMessageSize` and `max_message_size`.
See https://opendds.readthedocs.io/en/latest-release/devguide/run_time_configuration.html and https://youtu.be/-WzgzKrWmKA?t=2255.

**Reply by jeffj-archer** [reply]
Yes, this is using the built-in relay (RtpsRelay), sorry for not making that clear.

I will play around with max_message_size and SedpMaxMessageSize.  Does SedpMaxMessageSize only affect the discovery?  

What about max_packet_size?  I'm assuming we'd still want to take the DDS/OS default here, so as to be as efficient as possible.  The max_message_size will be the limiting factor as to how much data is getting pumped into any given packet.

For my test, yes, all payloads are 200K bytes.  That's my theoretical max...at present.

Appreciate your help, thank you!

**Reply by jrw972** [reply]
SedpMaxMessageSize only affects discovery because SEDP creates an rtps_udp transport for itself.  Thus, you need to configure both discovery and transport.  BTW, these settings should be common to the relay and the clients of the relay.

I don't think that `max_packet_size` is used by rtps_udp so the default is fine.

**Reply by jeffj-archer** [reply]
Would **max_samples_per_instance** be something of interest I should look into?  I don't want the reader blocking on a  certain # of samples to arrive before it consumes the sample (is a "sample" also a "message" ?).

**Reply by jrw972** [reply]
`max_samples_per_instance` is part of resource limits that cause old samples to be discarded when the limit is violated.
It's utility is determined by how you interact with the DataReader.
If you `take` samples, then it tends to not be useful because the number of samples stored by the reader tends to zero as the samples are taken.
If you `read` samples, then it may be useful to control how many samples the reader caches.

"Sample" is defined in the DDS spec.
DDS middleware like OpenDDS takes samples and use network messages to communicate them to other participants.
In general, a `write` will result in one or more outgoing messages assuming that there is at least one DataReader.
